### PR TITLE
IGNITE-21887 Sql. Reuse caller thread if query already prepared

### PR DESCRIPTION
**Pull request by korlov42**
https://issues.apache.org/jira/browse/IGNITE-21887

Before
|Benchmark|(clusterSize)|(fsync)|Mode|Cnt|Score|Error|Units|
|-----|-----|-----|-----|-----|-----|-----|-----|
|SelectBenchmark.sqlGet|1|false|avgt|20|55.288|± 3.919|us/op|
|InsertBenchmark.sqlInsert|1|false|avgt|20|0.437|± 0.018|ms/op|

After
|Benchmark|(clusterSize)|(fsync)|Mode|Cnt|Score|Error|Units|
|-----|-----|-----|-----|-----|-----|-----|-----|
|SelectBenchmark.sqlGet|1|false|avgt|20|10.265|± 0.137|us/op|
|InsertBenchmark.sqlInsert|1|false|avgt|20|0.367|± 0.017|ms/op|

The benchmarks were run on personal laptop(12th Gen Intel(R) Core(TM) i7-1280P, 32 GB RAM).

-----------------------------------

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

**Review by xtern — APPROVED** [review]

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java [hunk 1]
```diff
@@ -549,17 +550,6 @@ private CompletableFuture<QueryMetadata> prepareSingleAsync0(
             return prepareParsedStatement(schemaName, result, timestamp, queryCancel, params)
                     .thenApply(plan -> new QueryMetadata(plan.metadata(), plan.parameterMetadata()));
         });
-
-        // TODO IGNITE-20078 Improve (or remove) CancellationException handling.
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java:553]: should we close IGNITE-20078 ?

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/sql/ParserServiceImplTest.java [hunk 2]
```diff
@@ -49,54 +39,25 @@
...
-        ParsedResult expected = new DummyParsedResults();
-
-        ParsedResult actual = new ParserServiceImpl(0, new SameObjectCacheFactory(expected)).parse(statement.text);
-
-        assertSame(actual, expected);
-    }
-
-    @ParameterizedTest
-    @EnumSource(Statement.class)
-    void serviceCachesOnlyCertainStatements(Statement statement) {
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/sql/ParserServiceImplTest.java:81]: it looks like we are lost tests on caching

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java [hunk 3]
```diff
@@ -812,6 +793,20 @@ private static void validateDynamicParameters(int expectedParamsCount, Object[]
         }
     }
 
+    private static boolean shouldBeCached(SqlQueryType queryType) {
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java:796]: why shouldn't we cache EXPLAIN type?

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/sql/ParserServiceImplTest.java (hunk 2 above)
Comment (korlov42) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/sql/ParserServiceImplTest.java:81]: that is true. Unfortunately, SqlQueryProcessor is a mess right. now, it's hard to cover it with proper unit tests

Comment (ygerzhedovich) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/sql/ParserServiceImplTest.java:81]: could you please create a JIRA to cover it in the future

**Review by AMashenkov — APPROVED** [review]

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java (hunk 1 above)
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java:553]: yes, it looks not relevant anymore

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/sql/ParserServiceImplTest.java (hunk 2 above)
Comment (korlov42) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/sql/ParserServiceImplTest.java:81]: https://issues.apache.org/jira/browse/IGNITE-21932

**Review by ygerzhedovich — APPROVED** [review]

**Pull request merged**